os: add homedir()

cjihrig opened the pull request:
Do not land this yet.

This PR adds `os.homedir()`, which calls libuv's `uv_os_homedir()` to retrieve the current user's home directory. This is an alternative to #1670. This PR currently consists of two commits. 6887116da24dbd02f088318c84df63439f442c58 is the libuv update, which has not landed in io.js yet. This commit should be removed once a proper update is in place. e7e5992113a00e18005ef37d1eb820e5321c465b is the io.js update.

bnoordhuis commented on src/node_os.cc:
```diff
@@ -271,6 +271,29 @@ static void GetInterfaceAddresses(const FunctionCallbackInfo<Value>& args) {
+  char buf[MAX_PATH];
+#else
+  char buf[PATH_MAX];
+#endif
```
node.h defines PATH_MAX on Windows so the #ifdef is probably not necessary here.

bnoordhuis commented on src/node_os.cc:
```diff
@@ -271,6 +271,29 @@ static void GetInterfaceAddresses(const FunctionCallbackInfo<Value>& args) {
+#endif
+
+  size_t len = sizeof(buf);
+  int err = uv_os_homedir(buf, &len);
```
Maybe make the int const.

bnoordhuis commented on test/parallel/test-os.js:
```diff
@@ -101,3 +102,15 @@ switch (platform) {
+  assert.equal(home, process.env.USERPROFILE);
+} else if (!isWindows && process.env.HOME) {
+  assert.equal(home, process.env.HOME);
+}
```
Maybe as an extra check delete HOME / USERPROFILE from the env and check that `os.homedir()` still returns something.

cjihrig replied:
@bnoordhuis updated based on your feedback. I added a check that modifies the existing environment. I also added a new test that spawns a child process without the appropriate environment variables.

bnoordhuis commented on test/parallel/test-os.js:
```diff
@@ -101,3 +101,21 @@ switch (platform) {
+} else if (!common.isWindows && process.env.HOME) {
+  assert.equal(home, process.env.HOME);
+  delete process.env.HOME;
+  assert.equal(home, os.homedir());
```
I'm not sure if this is a good assumption.  I'd just check that it still returns _something_, possibly something that looks like a path.

cjihrig replied:
You're right. Updated.

bnoordhuis commented:
LGTM.  You should probably run the CI before landing this.

cjihrig commented:
I will once libuv is updated and I can remove the floating patch.

rvagg commented:
lgtm, the duplication in test-os.js seems unnecessary but I guess it's consistent with what's in there

Fishrock123 commented:
Libuv 1.6.0 has landed. :+1:

thefourtheye commented on deps/uv/src/unix/core.c:
```diff
@@ -983,3 +984,82 @@ int uv__dup2_cloexec(int oldfd, int newfd) {
+    return -EIO;
+
+  uid = getuid();
+  buf = NULL;
```
At this point, `buf` would be `NULL` only.

cjihrig commented:
Floating patch removed. CI: https://jenkins-iojs.nodesource.com/view/iojs/job/iojs+any-pr+multi/766/

cjihrig commented:
@rvagg the CI is still running, but I'm seeing some strange failures. Unable to compile on SmartOS, out of memory errors on FreeBSD, and too many files open on CentOS. Anything odd going on with the build?

rvagg commented:
@cjihrig eeek, not that I'm aware of, @jbergstroem has been tinkering on CentOS5 but not the others.

jbergstroem commented:
I'll have a look.

jbergstroem commented:
Affected os:es seems to hit file-max
```
freebsd: kern.maxfiles limit exceeded by uid 1003, please see tuning(7).
centos5: VFS: file-max limit 334442 reached
```

Jenkins has <100 file descriptors open which leads me to believe that ccache might try to. I can't say I've seen it previously, but I'll keep digging.

jbergstroem commented:
Looking at `cat /proc/sys/fs/file-nr` (or equivalents) at different machines; can't say I see it grow even close to limits while compiling.

jbergstroem commented:
(for reference, the run I'm measuring is here: https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/767/)

jbergstroem commented:
So yeah, I'm unable to reproduce what happened at that specific run. Jenkins.

cjihrig commented:
Added `limits.h` in  https://github.com/cjihrig/io.js/commit/69c1c33348ac1803615c73e42bf7dfef3155a4a7#diff-2b7b259d26f297c87b7d71512e33ccecR14 to get the compile working on SmartOS. The CentOS failure seemed to be flakey. FreeBSD is still throwing `ENOMEM`. @bnoordhuis any clue why this might be happening (other than the obvious answer that the process is out of memory)?

Updated CI run: https://jenkins-iojs.nodesource.com/view/iojs/job/iojs+any-pr+multi/769/

bnoordhuis commented:
I speculate it's because of (this)[https://github.com/freebsd/freebsd/blob/6fc6a3fe79eb196e331de929b9e74e40a6fb019f/lib/libc/gen/sysconf.c#L371-380]. :-)

(IIRC, I pointed that out in the review, that sysconf() can return < 0, but I'm not 100% sure.  At any rate, `bufsize` is a size_t so the -1 wraps around and becomes SIZE_MAX.)

cjihrig commented:
@bnoordhuis you did mention that. The data type for `bufsize` not being switched from `size_t` to `long` was my bad (I did add the check for `bufsize <= 0`, but didn't change the data type). Assuming that this is the problem, I'm curious why this didn't cause a problem with the libuv CI. And, do you think we should ditch the `sysconf()` call and use something else as the starting point for `malloc()`?

cjihrig commented:
@bnoordhuis you were right, so I added a37072b00d4b607bec6bdbcc06abd9f58d1b3e7d as a floating patch on libuv, and the [CI is much happier](https://jenkins-iojs.nodesource.com/view/iojs/job/iojs+any-pr+multi/770/) now. Let me know how you want this handled in libuv.

bnoordhuis commented:
@cjihrig Left a comment.  Can you file that as a libuv PR?  We'll do a patch release for that.

jbergstroem commented:
Ah, should've tested the branch while searching for issues; I just assumed it was lingering processes or similar.

cjihrig commented:
@bnoordhuis if I update the floating patch here, is this OK to land?

saghul commented:
@cjihrig FYI: https://github.com/nodejs/io.js/pull/1905

cjihrig commented:
Thanks @saghul. Hopefully one last CI run: https://jenkins-iojs.nodesource.com/view/iojs/job/iojs+any-pr+multi/774/

Fishrock123 commented:
Hmmm.. freebsd isn't so happy. https://jenkins-iojs.nodesource.com/view/iojs/job/iojs+any-pr+multi/774/nodes=freebsd101-64/tapTestReport/

cjihrig commented:
It would help if I rebased the branch to include the libuv 1.6.1 update.

https://jenkins-iojs.nodesource.com/view/iojs/job/iojs+any-pr+multi/775/ is all green.

Landed in 6e78e5feaa5a40888f1c49418f9c2c396838baa5.

sindresorhus commented:
I made a polyfill for anyone wanting to use this while still supporting older node/io versions:
https://github.com/sindresorhus/os-homedir
